### DISCUSSION TITLE
Constrained System.Type and dynamic generics

### DISCUSSION THREAD

**Discussion by fa1ntl1ght** [dotnet/csharplang#5267] [Language Ideas]
Update 11.11.2021: Better syntax.
`type` keyword replaced with `?` to be in line with java's wildcards.
`<>` replaced with preceding `*` due to relation with "indirection".

So, as static virtual methods are coming to C#, I have thought on some additional things that may improve their usability.
I don't know how to call this properly and if there have already been similar ideas (since I don't know what exactly to search for) and I haven't dug deep into possible complications that this idea can cause yet, so I'll just post my thoughts here and maybe someone who knows language design better can point me what may go wrong with it and why it's possibly a bad idea.

Now to the point.
In C# (and whole .NET) we represent types, that aren't exactly known at runtime, as `System.Type` and we use it to do a lot of things. For example, instantiating unknown types, calling methods on unknown types, accessing static fields and properties on unknown types and so on.
But what if the type isn't that unknown and we want to preserve type safety?
I imagined what it could look like if we had a possibility to look at `System.Type` objects as generic parameters.

Let's say we have these classes, the members of which we will access:
```cs
public class A
{
	public static virtual void SomeStaticMethod() => Console.WriteLine("A");
	
	public virtual void SomeInstanceMethod() => Console.WriteLine("a");
}

public class B : A
{
	public static override void SomeStaticMethod() => Console.WriteLine("B");
	
	public override void SomeInstanceMethod() => Console.WriteLine("b");
}
```
(Note that virtual static members in classes are _currently_ not supported in C#)

And we have a simple class, which will store a type, which can be either `A` or `B`, in field `someType`.
```cs
public class SomeClass
{
	public Type someType;
	
	public SomeClass(Type someType)
	{
		this.someType = someType;
	}
}
```

And lastly, we have a method, which is going to do some operations on a type from the field of received `SomeClass` object.
```cs
public static void Method(SomeClass cl)
{
/*
We are going to sequentially do the following actions:
- Print the name of the received type
- Call SomeStaticMethod on the type
- Create an instance of the type
- Call SomeInstanceMethod on the type
*/
	Console.WriteLine(cl.someType.Name);
	MethodInfo staticMethod = cl.someType.GetMethod(nameof(A.SomeInstanceMethod), BindingFlags.Public | BindingFlags.Instance | BindingFlags.FlattenHierarchy); //FlattenHierarchy is not actually needed there, but would be needed if there were a type which extends A, but doesn't override its methods
	staticMethod.Invoke(null, null);
	A obj = (A)Activator.CreateInstance(cl.someType);
	MethodInfo instanceMethod = cl.someType.GetMethod(nameof(A.SomeInstanceMethod), BindingFlags.Public | BindingFlags.Instance | BindingFlags.FlattenHierarchy);
	instanceMethod.Invoke(obj, null);
}
```
This example will work, provided static virtuals in classes are implemented in the language and I haven't made any syntax errors, but it is unnecessarily verbose and isn't type safe.
And my idea is to address these downsides by introducing a special syntax to describe constraints of `System.Type` symbols and treat these constrained types like generic arguments.
To illustrate what I mean, let's rewrite those `SomeClass` and `Method` with a new stub syntax.
```cs
public class SomeClass
{
	public Type<? : A, new()> someType; // Here, the Type is constrained to be any type derived from A with a 0-parameter constructor
	
	public SomeClass(Type<? : A, new()> someType)
	{
		this.someType = someType;
	}
}

public static void Method(SomeClass cl)
{
	Console.WriteLine(((Type)cl.someType).Name); //An explicit conversion is used to access System.Type members of cl.someType (without conversion we would be trying to access member Name of some type, constrained to be "? : A, new()")
	cl.someType.SomeStaticMethod(); //A direct call on constrained Type means access to a member of the type, that is represented by that Type object
	A obj = new *cl.someType(); //Here, an asterisk is used to distinguish the type, stored in cl.someType field, from literal cl.someType type that could also exist, despite it's against naming conventions
	obj.SomeInstanceMethod(); //A normal C# call on a class instance
}
```
So, to sum up, what that stub syntax does is:
- Allowing to use constrained `System.Type` type by using `Type<? : generic constraint>` syntax, which is similar to "where" clause of generic methods and classes, with the `?` here being a contextual keyword. During the compilation, this would be translated to regular System.Type type, and a special metadata attribute would be attached if this syntax comes up in an argument/field/property/return/etc type
- A reference to constrained type object is treated as a reference to some "virtual generic parameter" with the same constraint. So, if we have a variable `foo` of type `Type<? : A>` and we call `foo.Bar()` it is equivalent to a call on a generic parameter `<T> where T : A`, but instead of type argument, the source of the exact type is the underlying Type object in the variable
- An asterisk is used to distinguish references to constrained type variables from literal types. `new A<B>()` means creation of an object of type `A<B>`. `new A<*B>()` means creation of an object of type `A`, with type argument taken from a variable/field/property named `B` in the current scope.

That's probably all so far and I really want to get some opinion from the community and mostly I would like to know what I could have missed while trying to design this feature. I never done any language design and I understand that it may turn out that it's actually impossible to implement this feature.

**Comment by HaloFour** [comment]
Sounds like you want generics without the generics.  Why is it better to pass `Type` around here, even with some magic associated with it, over a generic type parameter?

**Reply by fa1ntl1ght** [reply]
> Sounds like you want generics without the generics. Why is it better to pass `Type` around here, even with some magic associated with it, over a generic type parameter?

Generic parameters are bound to types and methods and are visible from caller side, while Type can be obtained/(re)assigned/constructed/etc in the body of a method. If there were a way to get a constrained Type object from another object (like GetType(), but the returned Type would be constrained to be a type, derived from the type of the variable it was called on), then this constrained type object could be used to safely invoke static (virtual) methods on that type, call other generic methods, create new instances of generic classes with that type as argument without reflection and the obvious pros is the readability, provided there will be a good syntax.
An example of what I mean:
```cs
public class A { /* members...*/ }
public class B : A { /* members...*/ }
public class C : B { /* members...*/ }

public static void Method()
{
	A a = GetSomeA();
	Type<type : A> type = a.GetConstrainedType();
	IEnumerable<A> enumerable = new List<<type>>(); //and this list would have a's type as generic parameter
	SomeGenericMethod<<type>>();
	type.SomeVirtualStaticMethod();
}

public static void SomeGenericMethod<T>() where T : A { /* code... */ }
```
This feature can also probably go better with variance on classes

**Reply by HaloFour** [reply]
So this proposal is more about being able to use generics dynamically than any type variable, the ability to take the variable at runtime and use it as a generic type argument.  The runtime doesn't really offer any support for this outside of reflection.

**Reply by fa1ntl1ght** [reply]
> So this proposal is more about being able to use generics dynamically than any type variable, the ability to take the variable at runtime and use it as a generic type argument. The runtime doesn't really offer any support for this outside of reflection.

Yes, that's probably it. I don't know the exact limitations of the runtime, but you got my idea. Without runtime support this could be implemented as a compiler trick that translates down to reflection calls, but obviously runtime support is much better. In the worst case, at least the verbosity of reflection will be hidden and type safety will be present.

**Reply by 333fred** [reply]
Sounds like you want [associated/existential types](https://github.com/dotnet/csharplang/issues/1328).

**Reply by fa1ntl1ght** [reply]
> Sounds like you want [associated/existential types](https://github.com/dotnet/csharplang/issues/1328).

Thanks, I'll look at it

**Reply by Job-Bouwhuis** [reply]
in any case if you really NEED to use generic methods/types with Type variables you could always use Reflections to get the job done. 

like this
<img width="625" alt="Example Code" src="https://user-images.githubusercontent.com/92306873/213110580-bce8eb7e-0e43-4557-b131-3d473be90876.png">
